expression/constant_fold.go

```diff
@@ -17,25 +17,56 @@ import (
 	log "github.com/Sirupsen/logrus"
 )
 
-// FoldConstant does constant folding optimization on an expression.
+// FoldConstant does constant folding optimization on an expression excluding deferred ones.
 func FoldConstant(expr Expression) Expression {
-	scalarFunc, ok := expr.(*ScalarFunction)
-	if !ok || !scalarFunc.Function.canBeFolded() {
-		return expr
-	}
-	args := scalarFunc.GetArgs()
-	for i := 0; i < len(args); i++ {
-		if _, ok := args[i].(*Constant); !ok {
+	return foldConstant(expr, false)
+}
+
+// FoldDeferredConstant does constant folding optimization on an expression including deferred ones.
+func FoldDeferredConstant(expr Expression) Expression {
+	return foldConstant(expr, true)
+}
+
+func foldConstant(expr Expression, foldDeferred bool) Expression {
+	switch x := expr.(type) {
+	case *ScalarFunction:
+		if !x.Function.canBeFolded() {
 			return expr
 		}
+		args := x.GetArgs()
+		canFold := true
+		for i := 0; i < len(args); i++ {
+			foldedArg := foldConstant(args[i], foldDeferred)
+			x.GetArgs()[i] = foldedArg
+			con, conOK := foldedArg.(*Constant)
+			if !conOK || con.DeferredExpr != nil {
```

Comment:
The deferred constant should be able to fold.
If any of the argument is deferred constant, the result constant should set the `DefferredExpr` to the input `expr`.

Add a bool return value `isDeferredConstat` to `foldConstant` would be convenient.

Reply:
Thank you for the suggestions. I've updated the codes.